File: 0235-lowest-common-ancestor-of-a-binary-search-tree/0235-lowest-common-ancestor-of-a-binary-search-tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @param {TreeNode} p
# @param {TreeNode} q

# @return {TreeNode}
def lowest_common_ancestor(root, p, q)

    def lowest(root, p, q)
         # p root.val
        ret=0
         if root.left 
            res = lowest(root.left, p, q)
            return res if res.class == TreeNode
            ret+=res
         end
         if root.right
            res = lowest(root.right, p, q)
            return res if res.class == TreeNode
            ret+=res
         end
        
        ret+=1 if (root==p || root==q)
        return root if ret > 1
        ret
    end
    lowest(root, p, q)

end

File: 0235-lowest-common-ancestor-of-a-binary-search-tree/0235-lowest-common-ancestor-of-a-binary-search-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */

/**
 * @param {TreeNode} root
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {TreeNode}
 */
var lowestCommonAncestor = function(root, p, q) {
    
    
    // if right return and left return a different found 
    // if one is found and root is a match 
    let parent=null
    function findLCA(root){
        if(parent || !root)
            return null
        let left=findLCA(root.left)
        let right=findLCA(root.right)
        if(right&&left){
            parent=root
            return
        }
        else if(right||left){
            if(root.val==p.val || root.val==q.val){
                parent=root
                return 
            }
            return right || left
        }else if(root.val==p.val || root.val==q.val){
            return root
        }
    return null
    }
    findLCA(root)
    return parent 
};